ESQL: Push down `MvExpand` past `Project`

kanoshiou opened the pull request:
This PR enables `MV_EXPAND` to be pushed down past `Project` operations, reducing unnecessary field extractions caused by `KEEP *` and improving query efficiency.

Closes #136292
Closes #136596

I believe this PR also closes #119074. Although I couldn’t reproduce the plan mentioned there on current main, it addresses the same topic, so it should be safe to close that issue.

kanoshiou committed:
Push down MvExpand past Project

kanoshiou committed:
Update docs/changelog/136398.yaml

kanoshiou committed:
Merge remote-tracking branch 'origin/main' into push-down-mv_expand-past-project

kanoshiou committed:
Insert an eval if target in mv_expand is aliased in child project

kanoshiou committed:
Add tests for #136596

kanoshiou committed:
Update docs/changelog/136398.yaml

kanoshiou committed:
Merge branch 'refs/heads/main' into push-down-mv_expand-past-project

# Conflicts:
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java

kanoshiou committed:
Merge branch 'refs/heads/main' into push-down-mv_expand-past-project

# Conflicts:
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PushDownMvExpandPastProject.java:
```diff
@@ -0,0 +1,48 @@
+                        alias.child()
+                    );
+                    projections.set(i, mvExpand.expanded());
+                    pj = new Project(pj.source(), new Eval(aliasAlias.source(), pj.child(), List.of(aliasAlias)), projections);
```
I may be missing something, but I don't think we need to inject Evals. I expected we can just update the target field in the MvExpand to take the renames from the projection into account? (See comment above this one.)

alex-spies reviewed (changes requested):
Thanks a lot @kanoshiou !

I think this is very nice, but I think there's an approach that doesn't create defensive Evals to preserve the attribute. I'd like to explore that approach first before giving up on it.

kanoshiou committed:
Merge branch 'refs/heads/main' into push-down-mv_expand-past-project

# Conflicts:
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java

kanoshiou committed:
refactor(esql): Improve MvExpand push-down logic in logical optimizer

- Enhance PushDownMvExpandPastProject rule to handle more complex query scenarios
- Add checks to prevent duplicate output attributes when pushing down MvExpand
- Support pushing down MvExpand for queries with aliased expressions
- Handle cases where target is an alias or referenced in an alias
- Add test case to verify complex MvExpand push-down behavior
Resolves edge cases in multi-stage query optimization where MvExpand interactions with Project and Eval nodes were previously limited.

kanoshiou committed:
refactor(esql): Improve MvExpand push-down logic in logical optimizer

- Refactored PushDownMvExpandPastProject to handle more complex projection scenarios
- Added support for preserving original projection aliases during MvExpand push-down
- Replaced previous push-down implementation with more robust alias resolution
- Ensured that output expressions maintain their original context and aliases
- Updated test case to reflect new push-down behavior
The changes improve the flexibility and accuracy of MvExpand push-down optimization in the logical plan optimizer.

kanoshiou committed:
Merge branch 'refs/heads/main' into push-down-mv_expand-past-project

kanoshiou committed:
refactor(esql): Improve MvExpand push-down logic in logical optimizer

- Simplified push-down logic for MvExpand past Project operations
- Removed unnecessary Limit operation from example comment
- Used `replaceChild()` method instead of creating a new MvExpand instance
- Streamlined code for pushing down MvExpand past Project nodes
- Improved code readability and reduced complexity of optimizer rule

kanoshiou committed:
refactor(esql): Improve MvExpand push-down logic in logical optimizer

- Modify PushDownMvExpandPastProject to handle more complex projection scenarios
- Update PushDownUtils to make resolveRenamesFromProject method public
- Enhance test cases in LogicalPlanOptimizerTests to validate new push-down behavior
- Simplify MvExpand and Project interaction by removing unnecessary Eval steps
- Improve optimization logic for multi-stage MvExpand operations

kanoshiou committed:
Merge branch 'main' into push-down-mv_expand-past-project

kanoshiou commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PushDownMvExpandPastProject.java:
```diff
@@ -0,0 +1,48 @@
+            // Find if the target is aliased in the project and create an alias with temporary names for it.
+            for (int i = 0; i < projections.size(); i++) {
+                NamedExpression projection = projections.get(i);
+                if (projection instanceof Alias alias && alias.toAttribute().equals(mvExpand.target().toAttribute())) {
```
When the alias's original field is not referenced elsewhere in the projections, we don't need to inject an `Eval` node to preserve it. I've added a check to detect this scenario - if no other projection or alias references the same original field (the alias's child expression), we can safely call `resolveRenamesFromProject` directly and push down the `MvExpand` operation without the additional `Eval` overhead.

kanoshiou committed:
refactor(esql): Improve MvExpand push-down logic in PushDownMvExpandPastProject

- Refactor condition in semantic equality check for MvExpand push-down
- Modify logic to handle alias comparisons more precisely
- Simplify condition for checking input references in project operations
- Improve code readability and reduce complexity of push-down rule

luigidellaquila reviewed (commented):
I tried to wrap my head around this problem, and I came to the same conclusion as @kanoshiou: MV_EXPAND drops the source attribute, so we can't just move the aliasing after that, or we'll lose the original value.

This said, the combination of this new rule with `ReplaceAliasingEvalWithProject` tends to be confusing, because one reverts the other. Right now it's not problematic, because `ReplaceAliasingEvalWithProject` is in the Substitutions phase, while this new rule is in the Operator Optimizations phase, so they never run in a loop. If they did, we'd have an infinite loop.

Comments like [this](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java#L157) make me think that the problematic bit here is `ReplaceAliasingEvalWithProject`, and aliases in Project in general
(btw, we have at least another rule that puts aliases into projections that is `PushDownJoinPastProject`), but this makes the topic too complex probably.

On a side note, `ReplaceAliasingEvalWithProject` looks like a nice optimization, but I have no evidence that it is effective in practice (do we have a benchmark for it?). I guess an `EVAL a = b` could be easily optimized at compute level, just using a single block for both.

kanoshiou committed:
Merge branch 'refs/heads/main' into push-down-mv_expand-past-project

# Conflicts:
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java

alex-spies reviewed (commented):
Thanks @kanoshiou , you were right about the need for the defensive Evals.

I think this approach will work and have done another round of review. I think we can deal with the UnionAll problem, see below.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PushDownMvExpandPastProject.java:
```diff
@@ -0,0 +1,112 @@
+                        }
+
+                        // for query like: row a = 2 | eval b = a ｜ keep * | mv_expand b
+                        Alias aliasAlias = new Alias(
```
Why do we create a temporary alias for the target field?

I'd expect we can just take the already existing alias from the Project and leave it behind in the Eval.

In your example
```
row a = 2 | eval b = a ｜ keep * | mv_expand b
```
the alias from the `eval b = a` can just be left in place (except that it needs to be stuffed into the new eval rather than being part of the projections).

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PushDownMvExpandPastProject.java:
```diff
@@ -0,0 +1,112 @@
+            // Create a new projection at the top based on mvExpand.output(), plugging back in the aliases
+            List<NamedExpression> newProjections = new ArrayList<>();
+            for (Attribute outputExpr : mvExpand.output()) {
+                Alias alias = aliases.resolve(outputExpr, null);
```
We don't normally use `resolve` for an `AttributeMap<Alias>`. It's really meant for `AttributeMap<Attribute>` because it can deal with multiple indirections. We don't do that here, so we should rather just use `get`.

alex-spies reviewed (changes requested):
Sorry @kanoshiou , I found some things that don't make sense to me. Maybe I'm too tired and they will make more sense tomorrow, but I think we need to iterate one more time before we can get this in.

Let me know if you'd like to give it another shot - otherwise, I think I know how to touch up the optimizer rule and can happily get this over the finish line myself.

kanoshiou committed:
fix: refine `PushDownMvExpandPastProject` logic by correcting the alias collection and improving the pushdown skip condition.

kanoshiou committed:
Merge branch 'refs/heads/main' into push-down-mv_expand-past-project

kanoshiou committed:
fix(esql): simplify MvExpand pushdown logic by extracting to utility methods

- Extract MvExpand pushdown logic into reusable PushDownUtils methods
- Replace inline alias mapping with PushDownUtils.pushDownPastProject()
- Use PushDownUtils.resolveRenamesFromMap() to handle attribute resolution
- Reduce code duplication and improve maintainability of optimizer rules
- Simplify the projection reconstruction by delegating to utility functions

kanoshiou committed:
Update test

kanoshiou committed:
fix(esql): refactor MvExpand pushdown logic to directly manipulate projections

- Replace AttributeMap import with Attribute for more direct attribute handling
- Inline MvExpand pushdown logic instead of delegating to utility methods
- Directly update projections to point to expanded attributes during pushdown
- Handle both Alias and direct NamedExpression cases when replacing target references
- Simplify the overall flow by constructing Project directly with updated projections
- Improve code clarity and maintainability by making the transformation explicit

kanoshiou committed:
Avoid inconsistent plans

kanoshiou committed:
Create a temporary attribute for the UnionAll case

kanoshiou committed:
Add more tests for UnionAll cases

kanoshiou committed:
Merge branch 'refs/heads/main' into push-down-mv_expand-past-project

kanoshiou commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PushDownMvExpandPastProject.java:
```diff
@@ -0,0 +1,112 @@
+                            alias.child()
+                        );
+                        projections.set(i, mvExpand.expanded());
+                        pj = new Project(pj.source(), new Eval(aliasAlias.source(), pj.child(), List.of(aliasAlias)), projections);
```
Due to continuously encountering special cases while pushing down `mv_expand` past the `project`, I have always iterated on this branch based on the previous code versions. As a result, even I find this version of the branch quite confusing when trying to recall my original thought process. 

I have now entirely refactored the logic based on your suggestions, and I believe the new version is significantly cleaner and easier to understand.

kanoshiou committed:
Merge branch 'refs/heads/main' into push-down-mv_expand-past-project

# Conflicts:
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java

kanoshiou committed:
Merge branch 'refs/heads/main' into push-down-mv_expand-past-project

# Conflicts:
#	x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/generative/GenerativeRestTest.java

kanoshiou committed:
Merge branch 'refs/heads/main' into push-down-mv_expand-past-project

# Conflicts:
#	x-pack/plugin/esql/qa/testFixtures/src/main/resources/mv_expand.csv-spec

kanoshiou committed:
Merge branch 'refs/heads/main' into push-down-mv_expand-past-project

# Conflicts:
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java

kanoshiou committed:
Merge branch 'main' into push-down-mv_expand-past-project

# Conflicts:
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

wchaparro commented:
@alex-spies should we revive this one?

kanoshiou committed:
Merge branch 'main' into push-down-mv_expand-past-project

# Conflicts:
#	x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/generative/GenerativeRestTest.java
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java

kanoshiou committed:
fix test

astefan commented:
buildkite test this

kanoshiou committed:
checkstyle

kanoshiou committed:
fix test

kanoshiou committed:
Merge branch 'main' into push-down-mv_expand-past-project

kanoshiou committed:
fix tests

kanoshiou committed:
Merge branch 'main' into push-down-mv_expand-past-project

kanoshiou committed:
Merge branch 'refs/heads/main' into push-down-mv_expand-past-project

kanoshiou committed:
fix tests

kanoshiou committed:
Merge branch 'main' into push-down-mv_expand-past-project

kanoshiou committed:
fix: Refine `PushDownMvExpandPastProject` to correctly handle aliased projections and prevent duplicate output attributes when expanded field names conflict.

kanoshiou committed:
Merge branch 'main' into push-down-mv_expand-past-project

kanoshiou committed:
Merge branch 'refs/heads/main' into push-down-mv_expand-past-project

# Conflicts:
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizerTests.java
#	x-pack/plugin/esql/src/test/resources/org/elasticsearch/xpack/esql/plugin/golden_tests/LateMaterializationPlannerGoldenTests/testMvExpandBeforeTopN/local_reduce_physical_optimization_data_driver.expected
#	x-pack/plugin/esql/src/test/resources/org/elasticsearch/xpack/esql/plugin/golden_tests/LateMaterializationPlannerGoldenTests/testMvExpandBeforeTopN/local_reduce_planned_data_driver.expected
#	x-pack/plugin/esql/src/test/resources/org/elasticsearch/xpack/esql/plugin/golden_tests/LateMaterializationPlannerGoldenTests/testMvExpandBeforeTopN/local_reduce_planned_reduce_driver.expected
#	x-pack/plugin/esql/src/test/resources/org/elasticsearch/xpack/esql/plugin/golden_tests/LateMaterializationPlannerGoldenTests/testMvExpandBeforeTopN/physical_optimization.expected

kanoshiou committed:
Update test

kanoshiou committed:
Merge branch 'refs/heads/main' into push-down-mv_expand-past-project

# Conflicts:
#	x-pack/plugin/esql/src/test/resources/org/elasticsearch/xpack/esql/plugin/golden_tests/LateMaterializationPlannerGoldenTests/testMvExpandBeforeTopN/local_reduce_physical_optimization_data_driver.expected

kanoshiou committed:
update tests

kanoshiou committed:
Merge branch 'main' into push-down-mv_expand-past-project

# Conflicts:
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalLogicalPlanOptimizerTests.java

kanoshiou committed:
Merge branch 'main' into push-down-mv_expand-past-project

kanoshiou committed:
Update test

kanoshiou committed:
Merge branch 'refs/heads/main' into push-down-mv_expand-past-project

# Conflicts:
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalLogicalPlanOptimizerTests.java
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizerTests.java
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java

kanoshiou committed:
Merge branch 'main' into push-down-mv_expand-past-project

# Conflicts:
#	x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/generative/GenerativeRestTest.java

kanoshiou committed:
Merge branch 'refs/heads/main' into push-down-mv_expand-past-project

# Conflicts:
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java

kanoshiou committed:
Update testRedundantSortOnMvExpandJoinKeepDropRename

kanoshiou committed:
Merge branch 'main' into push-down-mv_expand-past-project

# Conflicts:
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java

wchaparro commented:
<!-- analytics-pr-review-management:stall-warn -->

Hi @kanoshiou — this PR has been identified as **stalled** (open and without meaningful updates for 30+ days).

Please address the requested changes when you can.

It will be **closed in two weeks** if there is no further activity. Please reply, push an update, or close it yourself if it is no longer needed.

— Analytics PR review management

kanoshiou committed:
Merge branch 'refs/heads/main' into push-down-mv_expand-past-project

# Conflicts:
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java

alex-spies commented:
Sorry @kanoshiou , the comment above is automated. I know you've already addressed the change request and this PR is awaiting review from us. I'm sorry it got stalled for so long; we had to deal with urgent stuff in the meantime and never got around to come back to this PR, even though it is definitely a useful change.

@wchaparro , this needs to be in our review queue, not auto-closed. The PR is old, but the issue is still worth addressing.

kanoshiou commented:
@alex-spies No worries, and thanks for the update! I appreciate you taking the time to revisit this PR. Glad to hear the change is still considered useful. Looking forward to the review!
